### PR TITLE
[Bugfix] Apply attention sinks in the Transformers backend

### PR DESCRIPTION
**Pull request by tdoublep**
## Purpose

vLLM PR #48270 adds native `GraniteSWAForCausalLM` / `GraniteMoeSWAForCausalLM` because "the transformers backend does NOT handle these models correctly. Sink tokens get dropped silently and so the model gives wrong output." That is accurate, and it is a bug in the Transformers modeling backend rather than something inherent to those models, so this PR fixes it there and every current and future sink model works on the backend.

Transformers models with learnable attention sinks (GPT-OSS, GraniteSWA, GraniteMoeSWA, DeepSeek V4, MiMo-V2-Flash, ...) hand the per-head sink to the attention interface as `s_aux`. Only the attention impl can fold a sink into the softmax denominator, so vLLM has to pass it to `Attention` at construction time. `vllm_attention_forward` accepted it in `**kwargs` and dropped it, so every softmax in the model was wrong.

Reproduced with `ibm-granite/granite-swash-2b` (greedy, transformers backend):

| prompt | before | after |
| --- | --- | --- |
| `The capital of France is` | `''` | `' Paris.\nThe capital of France is the largest city...'` |
| `def fibonacci(n):` | `'::::::::::::::::::::::::'` | a correct fibonacci body |
| `The boiling point of water at sea level is` | `'\n\n\n at at at at at'` | `' 100 degrees Celsius (212 degrees Fahrenheit).'` |

Two defects were involved:

1. **Sinks were dropped.** `Base.find_sinks` now locates the parameter Transformers keeps the sinks in, `Base.init_sink` materializes it as this rank's slice of the heads with a sharded weight loader, and `create_attention_instances` passes it as `Attention(sinks=...)`. `vllm_attention_forward` now raises instead of silently ignoring `s_aux` if a model applies sinks that we failed to find.
2. **`Attention` instances were invisible to `named_modules()`.** They are only stored in a plain dict, so `process_weights_after_loading` never ran for them, which is where FlashInfer upcasts sinks to the fp32 tensor its kernel requires. They are now registered in an `nn.ModuleList`, which also replaces the narrower `_vllm_mla_attn` workaround that existed for MLA layers.

## Test plan

New tests in `tests/models/transformers/test_backend.py`:

- `test_sinks` loads `tiny-random/gpt-oss-bf16` on the backend and asserts every attention layer has a sink and that the tensor it was handed matches the checkpoint. Skipped below SM90, where no backend supports sinks.
- `check_implementation` now also asserts that attention instances are registered submodules, covering defect 2 for every model it is used with.

Commands and results (1x GB200, transformers 5.15.0):

```
pytest tests/models/transformers/test_backend.py -k "sinks"               # 1 passed
pytest tests/models/transformers/test_backend.py -k "test_mla or hybrid"  # 2 passed
```

## Model evaluation

Prompt logprobs for a 55-token paragraph, `ibm-granite/granite-swash-2b`, bf16, against HF eager as the reference (mean logprob -1.5943):

| build | mean logprob | mean abs delta | max abs delta |
| --- | --- | --- | --- |
| main | -74.0868 | 72.4925 | 175.4590 |
| this PR | -7.8748 | 6.8942 | 81.0792 |

Greedy continuations become token-identical to HF with this PR. The residual logprob gap is a second, unrelated bug in the backend - Granite's `logits_scaling` divisor is ignored - fixed separately in #52158, which takes the same measurement to a mean absolute delta of 0.0165.

## Notes

- Not a duplicate: no open PR addresses sinks in the Transformers backend (checked `is:pr is:open` for "transformers backend" and "sinks"). #48270 adds native implementations for two Granite variants; this fixes the backend for all sink models. The two changes are compatible.
- Tensor parallel sink sharding follows the existing `sharded_weight_loader(0)` pattern (as in `gpt_oss.py`) but was not exercised, since only one GPU was available. Worth running `test_distributed`-style coverage before merge.
- AI assistance was used for this change (Claude Code).

### CODE REVIEW

File: vllm/model_executor/models/transformers/base.py [hunk 1]
```diff
@@ -172,6 +177,10 @@ def __init__(self, *, vllm_config: "VllmConfig", prefix: str = ""):
         self.recursive_replace()
         # Create attention instances for KV cache allocation
         self.attention_instances = self.create_attention_instances()
+        # The instances are passed to the model's forward through a plain dict, so
+        # register them as submodules too, otherwise `named_modules` never yields
+        # them and their `process_weights_after_loading` does not run
+        self._attention_layers = nn.ModuleList(self.attention_instances.values())
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/base.py:180]: I'm not sure if this is a problem for weight loading as the `kv_b_proj`/`sinks` won't have the right parents?

File: vllm/model_executor/models/transformers/__init__.py [hunk 2]
```diff
@@ -59,6 +59,13 @@ def vllm_attention_forward(
     self_attn = attention_instances[module.layer_idx]
     if scaling is not None:
         self_attn.impl.scale = float(scaling)
+    if kwargs.get("s_aux") is not None and not self_attn.has_sink:
+        raise ValueError(
+            f"{type(module).__name__} applies attention sinks, but the Transformers "
+            "modeling backend could not find the parameter holding them, so the "
+            "output would be wrong. Please open an issue at "
+            "https://github.com/vllm-project/vllm/issues/new"
+        )
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/__init__.py]: `s_aux` isn't actually used here so this just checks that the Transformers modelling code calls forward as expected. One way this could become load bearing would be if it validates that `s_aux == self_attn.sinks` in the first forward

**Comment by hmellor** [comment]
We should be able bolt this onto the new `AttentionFuser` that's being added in https://github.com/vllm-project/vllm/pull/54941.

It contains a very similar pattern where we want to read `scaling=` from the interface call and this PR can read `s_aux=`

**Comment by hmellor** [comment]
@claude review

**Comment by bohnstingl** [comment]
I can confirm that when running the BF16 GPT-OSS model (`lmsys/gpt-oss-20b-bf16`), I see sensible outputs with this fix.
For example, running the prompt
```bash
curl -s http://localhost:8000/v1/completions -H "Content-Type: application/json" -d '{"model":"lmsys/gpt-oss-20b-bf16","prompt":"The boiling point of water at sea level is","max_tokens":13,"temperature":0}'
```


| Branch | Output |
| -------- | -------- |
| main (`49eb2a`)   |  sea level is sea level is sea level is sea level is se |
| this branch    | 100°C (212°F). However, the boiling point of |

cc @tdoublep

**Review by hmellor — APPROVED** [review]
